---
title: R code to generate GMT of monocyte-derived vs tissue-resident macrophages
author: Slim Fourati
date: "17 April, 2019"
output: github_documents
---

Load required packages

```r
suppressPackageStartupMessages(library(package = "knitr"))
suppressPackageStartupMessages(library(package = "RCurl"))
suppressPackageStartupMessages(library(package = "readxl"))
suppressPackageStartupMessages(library(package = "biomaRt"))
suppressPackageStartupMessages(library(package = "GEOquery"))
suppressPackageStartupMessages(library(package = "DESeq2"))
suppressPackageStartupMessages(library(package = "tidyverse"))
```

Define session options

```r
workDir <- dirname(getwd())
opts_chunk$set(tidy = FALSE, fig.path = "../figure/")
options(stringsAsFactors  = FALSE,
	readr.num_columns = 0)
# create data.frame with genesets
gmtDF <- NULL
```

Dowload Misharin et al Supplementary Tables

```r
zipURL <- file.path("http://jem.rupress.org/highwire/filestream/128076",
		    "field_highwire_adjunct_files/0",
		    "JEM_20162152_TablesS1-S6.zip")
zipBin <- getBinaryURL(url = zipURL, followlocation = TRUE)
zipFile <- basename(zipURL)
zipCon <- file(zipFile, open = "wb")
writeBin(zipBin, zipCon)
close(zipCon)
```

Extract list of genes induced in Monocytes derived macrophage compared to 
tissue-residents macrophages (cluster III)

```r
zipDir <- unzip(zipfile = zipFile)
tableS2 <- grep(pattern = "TableS2", zipDir, value = TRUE) %>%
  read_excel(.name_repair = make.unique)
# extract gene_id
cluster3IDs <- tableS2 %>%
  filter(`K-assignment` %in% 3 & Anova_FDR_adj_Qvalue <= 0.05) %>%
  .$Symbol
# clean up
unlink(zipFile)
unlink(unique(dirname(zipDir)), recursive = TRUE)
```

Convert mouse genes to human genes

```r
human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

mouse2human <- getLDS(attributes  = c("ensembl_gene_id", "mgi_symbol"),
		      filters     = "ensembl_gene_id",
		      values      = cluster3IDs,
		      mart        = mouse,
		      attributesL = "hgnc_symbol",
		      martL       = human,
		      uniqueRows  = TRUE)
```

Create geneset based on Misharin et al cluster III genes

```r
gmtDF <- mouse2human %>%
  rowid_to_column() %>%
  select(rowid, HGNC.symbol) %>%
  spread(rowid, HGNC.symbol) %>%
  mutate(NAME = "MISHARIN_CLUSTER_III",
         DESC = paste("Genes incuded in monocyte-derived macrophages",
                      "compared to tissue resident-macrophages and blood",
                      "monocytes")) %>%
  select(NAME, DESC, setdiff(names(.), c("NAME", "DESC")))
```



Read Mould count file

```r
supFileLS <- getGEOSuppFiles(GEO           = "GSE94749",
			     filter_regex  = "htseq-genecounts.txt",
			     makeDirectory = FALSE) %>%
  rownames_to_column()
countFile <- basename(supFileLS$rowname)

countDF <- read_tsv(file = countFile) %>%
  as.data.frame() %>%
  column_to_rownames(var = "GeneID")

annotDF <- data.frame(condition = gsub(pattern = "_rep[0-9]",
                                       replacement = "",
                                       names(countDF)),
		      rowname   = names(countDF)) %>%
  mutate(condition = factor(condition)) %>%
  column_to_rownames(var = "rowname")
# clean up
unlink(countFile)
```



Perform DEG analysis with DESeq2 and identify genes differentially expressed
with abs(FC)≥2 and FDR≤0.1

```r
dds <- DESeqDataSetFromMatrix(countData = countDF,
                              colData   = annotDF,
                              design    = ~ 0 + condition)
dds <- DESeq(dds)
degLS <- list()
recLS <- grep(pattern = "Rec", levels(colData(dds)$condition), value = TRUE)
for (COND in recLS) {
  resCond <- gsub(pattern = "Rec", replacement = "Res", COND)
  res <- results(dds, contrast = c("condition",
                                   COND,
				   resCond))
  top <- res %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    filter(abs(log2FoldChange) >= 1 & padj <= 0.1)
  degLS[[toupper(paste0("MOULD_", COND, "_VS_", resCond, "_UP"))]] <- top %>%
    filter(log2FoldChange > 0) %>%
    .$rowname
  degLS[[toupper(paste0("MOULD_", COND, "_VS_", resCond, "_TOP200_UP"))]] <- top %>%
	  filter(log2FoldChange > 0) %>%
	  arrange(desc(stat)) %>%
	  top_n(n = 200, wt = stat) %>%
	  .$rowname
  degLS[[toupper(paste0("MOULD_", COND, "_VS_", resCond, "_DN"))]] <- top %>%
	  filter(log2FoldChange < 0) %>%
	  .$rowname
  degLS[[toupper(paste0("MOULD_", COND, "_VS_", resCond, "_TOP200_DN"))]] <- top %>%
    filter(log2FoldChange < 0) %>%
    arrange(stat) %>%
    top_n(n = 200, wt = -stat) %>%
    .$rowname
}
```

Convert geneids to human gene symbols

```r
degDF <- stack(degLS) %>%
  mutate(ind = as.vector(ind))

mouse2human <- getLDS(attributes  = c("ensembl_gene_id", "mgi_symbol"),
		      filters     = "ensembl_gene_id",
		      values      = unique(degDF$values),
		      mart        = mouse,
		      attributesL = "hgnc_symbol",
		      martL       = human,
		      uniqueRows  = TRUE)

geneLS <- merge(x = degDF,
		y = mouse2human,
		by.x = "values",
		by.y = "Gene.stable.ID") %>%
  filter(!is.na(HGNC.symbol)) %>%
  select(HGNC.symbol, ind) %>%
  distinct() %>%
  unstack()
```

Write GMT

```r
# write Misharin GMT
write_tsv(gmtDF,
          path      = file.path(workDir,
				"utils/Macrophage.gmt"),
          col_names = FALSE)
# write Mould GMT
for (i in 1:length(geneLS)) {
  geneLS[i] %>%
    stack() %>%
    rowid_to_column() %>%
    spread(rowid, values) %>%
    mutate(DESC = ind) %>%
    select(ind, DESC, setdiff(names(.), c("ind", "DESC"))) %>%
    write_tsv(path      = file.path(workDir,
				    "utils/Macrophage.gmt"),
	      col_names = FALSE,
	      append    = TRUE)
}
```

Print session info

```r
sessionInfo()
```

```
## R version 3.5.3 (2019-03-11)
## Platform: x86_64-apple-darwin18.2.0 (64-bit)
## Running under: macOS Mojave 10.14.4
## 
## Matrix products: default
## BLAS/LAPACK: /usr/local/Cellar/openblas/0.3.5/lib/libopenblasp-r0.3.5.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] forcats_0.4.0               stringr_1.4.0              
##  [3] dplyr_0.8.0.1               purrr_0.3.2                
##  [5] readr_1.3.1                 tidyr_0.8.3                
##  [7] tibble_2.1.1                ggplot2_3.1.0              
##  [9] tidyverse_1.2.1             DESeq2_1.22.2              
## [11] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
## [13] BiocParallel_1.16.6         matrixStats_0.54.0         
## [15] GenomicRanges_1.34.0        GenomeInfoDb_1.18.2        
## [17] IRanges_2.16.0              S4Vectors_0.20.1           
## [19] GEOquery_2.50.5             Biobase_2.42.0             
## [21] BiocGenerics_0.28.0         biomaRt_2.38.0             
## [23] readxl_1.3.1                RCurl_1.95-4.12            
## [25] bitops_1.0-6                knitr_1.22                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-137           lubridate_1.7.4        bit64_0.9-7           
##  [4] RColorBrewer_1.1-2     progress_1.2.0         httr_1.4.0            
##  [7] tools_3.5.3            backports_1.1.3        R6_2.4.0              
## [10] rpart_4.1-13           Hmisc_4.2-0            DBI_1.0.0             
## [13] lazyeval_0.2.2         colorspace_1.4-1       nnet_7.3-12           
## [16] withr_2.1.2            tidyselect_0.2.5       gridExtra_2.3         
## [19] prettyunits_1.0.2      curl_3.3               bit_1.1-14            
## [22] compiler_3.5.3         cli_1.1.0              rvest_0.3.2           
## [25] htmlTable_1.13.1       xml2_1.2.0             scales_1.0.0          
## [28] checkmate_1.9.1        genefilter_1.64.0      digest_0.6.18         
## [31] foreign_0.8-71         XVector_0.22.0         base64enc_0.1-3       
## [34] pkgconfig_2.0.2        htmltools_0.3.6        limma_3.38.3          
## [37] htmlwidgets_1.3        rlang_0.3.1            rstudioapi_0.9.0      
## [40] RSQLite_2.1.1          generics_0.0.2         jsonlite_1.6          
## [43] acepack_1.4.1          magrittr_1.5           GenomeInfoDbData_1.2.0
## [46] Formula_1.2-3          Matrix_1.2-15          Rcpp_1.0.1            
## [49] munsell_0.5.0          stringi_1.4.3          zlibbioc_1.28.0       
## [52] plyr_1.8.4             grid_3.5.3             blob_1.1.1            
## [55] crayon_1.3.4           lattice_0.20-38        haven_2.1.0           
## [58] splines_3.5.3          annotate_1.60.1        hms_0.4.2             
## [61] locfit_1.5-9.1         pillar_1.3.1           geneplotter_1.60.0    
## [64] XML_3.98-1.19          glue_1.3.1             evaluate_0.13         
## [67] latticeExtra_0.6-28    modelr_0.1.4           data.table_1.12.0     
## [70] cellranger_1.1.0       gtable_0.2.0           assertthat_0.2.0      
## [73] xfun_0.5               broom_0.5.1            xtable_1.8-3          
## [76] survival_2.43-3        AnnotationDbi_1.44.0   memoise_1.1.0         
## [79] cluster_2.0.7-1
```
