---
title: FluOMICs lung gene-level analysis
author: Slim Fourati
date: "06 April, 2019"
output: github_documents
---

Load packages

```r
suppressPackageStartupMessages(library(package = "knitr"))
suppressPackageStartupMessages(library(package = "tidyverse"))

suppressPackageStartupMessages(library(package = "edgeR"))
suppressPackageStartupMessages(library(package = "Biobase"))
suppressPackageStartupMessages(library(package = "pheatmap"))
suppressPackageStartupMessages(library(package = "biomaRt"))
suppressPackageStartupMessages(library(package = "impute"))
suppressPackageStartupMessages(library(package = "mixOmics"))
suppressPackageStartupMessages(library(package = "igraph"))
suppressPackageStartupMessages(library(package = "readxl"))
suppressPackageStartupMessages(library(package = "org.Mm.eg.db"))
suppressPackageStartupMessages(library(package = "glmnet"))
suppressPackageStartupMessages(library(package = "RCurl"))
suppressPackageStartupMessages(library(package = "jsonlite"))
```

Set session options

```r
workDir <- dirname(getwd())
opts_chunk$set(tidy = FALSE, fig.path = "../figure/")
options(readr.num_columns = 0,
        stringsAsFactors  = FALSE)
```

Read raw counts


Read sample annotation

```r
sampleAnnotFile  <- file.path(workDir,
			      "input/SraRunTable.txt")
sampleAnnotation <- read_tsv(file = sampleAnnotFile)
```

Read gene annotation

```r
cNames <- c("seqname",
	    "source",
	    "feature",
	    "start",
	    "end",
	    "score",
	    "strand",
	    "frame",
	    "attribute")
colTypes <- paste(c(rep("c", times = 3),
                    rep("i", times = 2),
                    rep("c", times = 4)),
                  collapse = "")
featuresAnnotationFile <- "genes.gtf"
skipNb <- read_lines(file = file.path(workDir, "input", featuresAnnotationFile))
skipNb <- sum(grepl(pattern = "^#", skipNb))
```





exploratory analysis: MDS plot


Differential expression analysis




plot heatmap with top50 differentially expressed genes

