---
title: FluOMICs lung gene-level analysis
author: Slim Fourati
date: "08 April, 2019"
output: github_documents
---

Load packages

```r
suppressPackageStartupMessages(library(package = "knitr"))
suppressPackageStartupMessages(library(package = "tidyverse"))
suppressPackageStartupMessages(library(package = "EDASeq"))
suppressPackageStartupMessages(library(package = "edgeR"))
#suppressPackageStartupMessages(library(package = "Biobase"))
#suppressPackageStartupMessages(library(package = "pheatmap"))
#suppressPackageStartupMessages(library(package = "biomaRt"))
#suppressPackageStartupMessages(library(package = "impute"))
#suppressPackageStartupMessages(library(package = "mixOmics"))
#suppressPackageStartupMessages(library(package = "igraph"))
#suppressPackageStartupMessages(library(package = "readxl"))
#suppressPackageStartupMessages(library(package = "org.Mm.eg.db"))
#suppressPackageStartupMessages(library(package = "glmnet"))
#suppressPackageStartupMessages(library(package = "RCurl"))
#suppressPackageStartupMessages(library(package = "jsonlite"))
```

Set session options

```r
workDir <- dirname(getwd())
opts_chunk$set(tidy = FALSE, fig.path = "../figure/")
options(readr.num_columns = 0,
        stringsAsFactors  = FALSE)
```

Read raw counts

```r
seqFile <- file.path(workDir, "input/lung.genecounts.csv")
countMat <- read_csv(file = seqFile) %>%
  as.data.frame() %>%
  column_to_rownames(var = "gene_id") %>%
  as.matrix()
cat("First five rows/columns of count matrix")
```

```
## First five rows/columns of count matrix
```

```r
print(countMat[1:5, 1:5])
```

```
##                    SRR5506389 SRR5506390 SRR5506391 SRR5506392 SRR5506393
## ENSMUSG00000000001       6887       5892       6360       2215       2170
## ENSMUSG00000000003          0          0          0          0          0
## ENSMUSG00000000028        131        124        145         49         33
## ENSMUSG00000000031       2514       1735       1341       1060        414
## ENSMUSG00000000037         20         15         17         11          7
```

```r
cat("\nDimensions of the count matrix")
```

```
## 
## Dimensions of the count matrix
```

```r
print(dim(countMat))
```

```
## [1] 53465    48
```

Read sample annotation

```r
sampleAnnot <- read_tsv(file = file.path(workDir,
					 "input",
					 "SraRunTable.txt")) %>%
  mutate(virus = gsub(pattern = "^.+\\(([^ |^ \\)]+).*$",
		      replacement = "\\1",
		      infection),
	 rowname = Run) %>%
  column_to_rownames(var = "rowname")
sampleAnnot <- sampleAnnot[colnames(countMat), ]
```

Read gene annotation

```r
colNames <- c("seqname",
	    "source",
	    "feature",
	    "start",
	    "end",
	    "score",
	    "strand",
	    "frame",
	    "attribute")
colTypes <- paste(c(rep("c", times = 3),
                    rep("i", times = 2),
                    rep("c", times = 4)),
                  collapse = "")
featuresAnnotationFile <- "genes.gtf"
skipNb <- read_lines(file = file.path(workDir, "input", featuresAnnotationFile))
skipNb <- sum(grepl(pattern = "^#", skipNb))

featuresAnnot <- read_tsv(file      = file.path(workDir,
						"input",
                                                featuresAnnotationFile),
                          skip      = skipNb,
                          col_names = colNames,
                          col_types = colTypes)
# extract gene_id and transcript_id from attributes
featAnnot <- featuresAnnot %>%
  mutate(gene_id = gsub(pattern = ".*gene_id \"([^;]+)\";.+",
                        replacement = "\\1",
                        attribute),
         gene_name = ifelse(test = grepl(pattern = "gene_name",
                                         attribute),
                            yes = gsub(pattern = ".+gene_name \"([^;]+)\";.+",
                                       replacement = "\\1",
                                       attribute),
                            no  = NA),
         gene_biotype = ifelse(test = grepl(pattern = "gene_biotype",
                                         attribute),
                            yes = gsub(pattern = ".+gene_biotype \"([^;]+)\";.*",
                                       replacement = "\\1",
                                       attribute),
                            no  = NA)) %>%
  select(seqname, strand, gene_id, gene_name, gene_biotype) %>%
  distinct() %>%
  as.data.frame()
rownames(featAnnot) <- featAnnot$gene_id
featAnnot <- featAnnot[rownames(countMat), ]
cat("Number of genes annotated")
```

```
## Number of genes annotated
```

```r
featAnnot %>%
    group_by(gene_biotype) %>%
    summarize(n = n()) %>%
    arrange(desc(n))
```

```
## # A tibble: 44 x 2
##    gene_biotype               n
##    <chr>                  <int>
##  1 protein_coding         22040
##  2 processed_pseudogene    8996
##  3 lincRNA                 5307
##  4 TEC                     3088
##  5 antisense_RNA           2727
##  6 unprocessed_pseudogene  2568
##  7 miRNA                   2207
##  8 snoRNA                  1507
##  9 snRNA                   1385
## 10 processed_transcript     777
## # â€¦ with 34 more rows
```


```r
seqSetRaw <- newSeqExpressionSet(counts      = countMat,
				 phenoData   = sampleAnnot,
				 featureData = featAnnot)
save(seqSetRaw, file = file.path(workDir, "output/fluomics.seqSetRaw.RData"))
```


```r
dge <- DGEList(counts       = counts(seqSetRaw),
               remove.zeros = TRUE)
```

```
## Removing 15376 rows with all zero counts
```

```r
dge <- calcNormFactors(object = dge, method = "TMM")
normMat <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE)
rawMat <- counts(seqSetRaw)[rownames(normMat), colnames(normMat)]
sampleAnnot <- pData(seqSetRaw)[colnames(normMat), ]
featAnnot <- fData(seqSetRaw)[rownames(normMat), ]
seqSet <- newSeqExpressionSet(
  counts           = rawMat,
  normalizedCounts = normMat,
  featureData      = featAnnot,
  phenoData        = sampleAnnot)
save(seqSet, file = file.path(workDir, "output/fluomics.seqSet.RData"))
```


```r
# baselined on mock
normMat0 <- normCounts(seqSet)[, seqSet$virus %in% "mock"] %>%
  rowMeans()
seqSetBaselined <- seqSet
normCounts(seqSetBaselined) <- normCounts(seqSet) - normMat0
seqSetBaselined <- seqSetBaselined[, seqSetBaselined$virus != "mock"]
save(seqSetBaselined, file = file.path(workDir, "output/fluomics.seqSetBaselined.RData"))
```



exploratory analysis: MDS plot

```r
countMat <- normCounts(seqSet)
distMat <- dist(t(countMat))
mds <- cmdscale(d = distMat, k = ncol(countMat) - 1, eig = TRUE)
mdsStDev <- apply(mds$points, MARGIN = 2, FUN = sd)
mdsPercVar <- round((mdsStDev^2)/sum(mdsStDev^2) * 100)
plotDF <- data.frame(V1 = mds$points[, 1],
                     V2 = mds$points[, 2]) %>%
  cbind(pData(seqSet))
ggplot(data = plotDF,
       mapping = aes(x = V1, y = V2, color = time_point, shape = virus)) +
  geom_point(size = 3) +
  labs(x     = paste0("1st dimension (", mdsPercVar[1], "%)"),
       y     = paste0("2nd dimension (", mdsPercVar[2], "%)"),
       title = "Multidimentional scaling plot") +
  theme_bw() +
  theme(legend.key = element_blank())
```

Differential expression analysis

```r
fits <- list()
virusTime <- interaction(seqSet$virus, seqSet$time_point, drop = TRUE)
designMat <- model.matrix(~0+virusTime)
colnames(designMat) <- gsub(pattern     = "virusTime",
			    replacement = "",
			    colnames(designMat))
rownames(designMat) <- sampleNames(seqSet)
dge <- DGEList(counts = counts(seqSet))
dge <- calcNormFactors(object = dge, method = "TMM")
dge <- estimateGLMCommonDisp(y = dge, design = designMat)
dge <- estimateGLMTrendedDisp(y = dge, design = designMat)
dge <- estimateGLMTagwiseDisp(y = dge, design = designMat)
fit <- glmFit(dge, design = designMat)
contrastLS <- grep(pattern = "^H5N1", colnames(designMat), value = TRUE) %>%
  data.frame(coef1 = .) %>%
  mutate(coef2 = gsub(pattern = "H5N1", replacement = "H1N1", coef1),
         contrast = paste0(coef1, "-", coef2)) %>%
  .$contrast
contrastMat <- makeContrasts(contrasts = contrastLS, levels = designMat)
fit$contrast <- contrastMat
fits[["virus"]] <- list(fit = fit)
# infection compared to mock
contrastLS <- setdiff(colnames(designMat), "mock.04d") %>%
  paste0("-mock.04d")
contrastMat <- makeContrasts(contrasts = contrastLS, levels = designMat)
fit$contrast <- contrastMat
fits[["infection"]] <- list(fit = fit)
save(fits, file = file.path(workDir, "output/fluomics.fits.RData"))
```


```r
load(file = "data/fits.fluomics.RData")
# print number of DEG
topDF <- NULL
for (modelName in names(fits)) {
  fit <- fits[[modelName]][["fit"]]
  for (i in 1:ncol(fit$contrast)) {
    top <- glmLRT(glmfit = fit, contrast = fit$contrast[, i]) %>%
      topTags(n = Inf) %>%
      as.data.frame() %>%
      group_by(sign(logFC)) %>%
      summarize(adj.p = sum(FDR <= 0.05),
                p     = sum(PValue <= 0.05)) %>%
      mutate(modelName = modelName,
             coef = colnames(fit$contrast)[i])
    topDF <- rbind(topDF, top)
  }
}
topDF %>%
  filter(`sign(logFC)` != 0) %>%
  gather(cname, value, -`sign(logFC)`, -modelName, -coef) %>%
  spread(`sign(logFC)`, value) %>%
  mutate(value = paste0(`1`, "/", `-1`)) %>%
  select(-`1`, -`-1`) %>%
  spread(cname, value) %>%
  as.data.frame() %>%
  print()
```

plot heatmap with top50 differentially expressed genes

```r
for (i in 1:ncol(fit$contrast)) {
  top <- glmLRT(glmfit = fit, contrast = fit$contrast[, i]) %>%
    topTags(n = 50) %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    arrange(sign(logFC), PValue)

  condLS <- which(fit$contrast[, i] != 0)
  sampleLS <- fit$design[, condLS] %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    gather(cond, value, -rowname) %>%
    filter(value != 0)
  mat <- exprs(seqSetBaselined)[top$rowname, sampleLS$rowname] %>%
    pmax(0.1) %>%
    log2()
  colorLS <- colorRampPalette(colors = c("blue", "white", "red"))(100)
  breakLS <- c(-1 * max(abs(mat)),
               seq(from = -1 * min(abs(range(mat))),
                   to   = min(abs(range(mat))),
                   length.out = 99),
               max(abs(mat)))
  rowAnnot <- top %>%
    mutate(sign = sign(logFC),
           sign = factor(sign)) %>%
    select(sign)
  rownames(rowAnnot) <- top$rowname
  pheatmap(mat = mat,
           cluster_cols = FALSE,
           cluster_rows = FALSE,
           colors = colorLS,
           breaks = breakLS,
           cellwidth = 10,
           cellheight= 5,
           fontsize_row = 5,
           main = colnames(fit$contrast)[i],
           annotation_row = rowAnnot)
}
```

Print session info

```r
sessionInfo()
```

```
## R version 3.5.3 (2019-03-11)
## Platform: x86_64-apple-darwin18.2.0 (64-bit)
## Running under: macOS Mojave 10.14.4
## 
## Matrix products: default
## BLAS/LAPACK: /usr/local/Cellar/openblas/0.3.5/lib/libopenblasp-r0.3.5.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] edgeR_3.24.3                limma_3.38.3               
##  [3] EDASeq_2.16.3               ShortRead_1.40.0           
##  [5] GenomicAlignments_1.18.1    SummarizedExperiment_1.12.0
##  [7] DelayedArray_0.8.0          matrixStats_0.54.0         
##  [9] Rsamtools_1.34.1            GenomicRanges_1.34.0       
## [11] GenomeInfoDb_1.18.2         Biostrings_2.50.2          
## [13] XVector_0.22.0              IRanges_2.16.0             
## [15] S4Vectors_0.20.1            BiocParallel_1.16.6        
## [17] Biobase_2.42.0              BiocGenerics_0.28.0        
## [19] forcats_0.4.0               stringr_1.4.0              
## [21] dplyr_0.8.0.1               purrr_0.3.2                
## [23] readr_1.3.1                 tidyr_0.8.3                
## [25] tibble_2.1.1                ggplot2_3.1.0              
## [27] tidyverse_1.2.1             knitr_1.22                 
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-137           bitops_1.0-6           lubridate_1.7.4       
##  [4] bit64_0.9-7            RColorBrewer_1.1-2     progress_1.2.0        
##  [7] httr_1.4.0             tools_3.5.3            backports_1.1.3       
## [10] utf8_1.1.4             R6_2.4.0               DBI_1.0.0             
## [13] lazyeval_0.2.2         colorspace_1.4-1       withr_2.1.2           
## [16] prettyunits_1.0.2      tidyselect_0.2.5       bit_1.1-14            
## [19] compiler_3.5.3         cli_1.1.0              rvest_0.3.2           
## [22] xml2_1.2.0             rtracklayer_1.42.2     scales_1.0.0          
## [25] genefilter_1.64.0      DESeq_1.34.1           digest_0.6.18         
## [28] R.utils_2.8.0          pkgconfig_2.0.2        rlang_0.3.1           
## [31] readxl_1.3.1           rstudioapi_0.9.0       RSQLite_2.1.1         
## [34] generics_0.0.2         hwriter_1.3.2          jsonlite_1.6          
## [37] R.oo_1.22.0            RCurl_1.95-4.12        magrittr_1.5          
## [40] GenomeInfoDbData_1.2.0 Matrix_1.2-15          fansi_0.4.0           
## [43] Rcpp_1.0.1             munsell_0.5.0          R.methodsS3_1.7.1     
## [46] stringi_1.4.3          zlibbioc_1.28.0        plyr_1.8.4            
## [49] grid_3.5.3             blob_1.1.1             crayon_1.3.4          
## [52] lattice_0.20-38        haven_2.1.0            splines_3.5.3         
## [55] GenomicFeatures_1.34.6 annotate_1.60.1        hms_0.4.2             
## [58] locfit_1.5-9.1         pillar_1.3.1           geneplotter_1.60.0    
## [61] biomaRt_2.38.0         XML_3.98-1.19          glue_1.3.1            
## [64] evaluate_0.13          latticeExtra_0.6-28    modelr_0.1.4          
## [67] cellranger_1.1.0       gtable_0.2.0           assertthat_0.2.0      
## [70] xfun_0.5               aroma.light_3.12.0     xtable_1.8-3          
## [73] broom_0.5.1            survival_2.43-3        AnnotationDbi_1.44.0  
## [76] memoise_1.1.0
```
